### PR TITLE
feat(manager/pip-compile): Use python keychain module for credentials

### PR DESCRIPTION
**Pull request by mbudnek**
## Changes

This refactors the pip-compile manager to use the python keychain module along with the keychains.envvars module to pass credentials to pip. Doing it this way means there's a significantly lower likelyhood of accidentally leaking credentials into the output file, since credentials looked up via the keychain module won't ever be written to --index-url or --extra-index-url directives in the output file.

Because of this lower likelihood of leaking credentials, I relaxed the usage of --no-emit-index-url to only automatically apply when credentials are found in index URLs in environment variables.

## Context

https://github.com/renovatebot/renovate/pull/26853#discussion_r1480524305
https://github.com/containerbase/base/pull/2145

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/pip-compile/artifacts.spec.ts [hunk 1]
```diff
@@ -345,21 +351,31 @@ describe('modules/manager/pip-compile/artifacts', () => {
...
+    const ORIGINAL_ENV = process.env;
+
+    beforeEach(() => {
+      jest.resetModules();
+      process.env = { ...ORIGINAL_ENV };
+    });
+
+    afterAll(() => {
+      process.env = ORIGINAL_ENV;
+    });
```
Comment (viceice) [inline lib/modules/manager/pip-compile/artifacts.spec.ts:370]: please avoid this, just delete unwanted ENV entries on before each

File: lib/modules/manager/pip-compile/common.ts [hunk 2]
```diff
@@ -212,71 +215,46 @@ function throwForUnknownOption(arg: string): void {
...
-  return {
-    haveCredentials: indexUrl.haveCredentials || extraIndexUrls.haveCredentials,
-    environmentVars: {
-      ...indexUrl.environmentVars,
-      ...extraIndexUrls.environmentVars,
-    },
-  };
+  const ret = [...uniqueHosts]
+    .map(getRegistryCredEnvVars)
+    .reduce((allCreds, current) => Object.assign(allCreds, current), {});
```
Comment (viceice) [inline lib/modules/manager/pip-compile/common.ts:257]: please use a for-of loop

**Review by viceice — APPROVED** [review]
changed to feat, so we get a feature release because it chnages behavior a lot.

a refactor PR should never change behavior.

**Review by secustor — APPROVED** [review]

**Review by not7cd — APPROVED** [review]
I will evaluate this feature after release.

I would like to see mention of `--no-emit-index-url` behaviour and link to _private-packages.md_ in pip-compile _readme.md_. Manager specific documentation is often the first thing someone sees.

**Pull request merged**